torch/functional.py

```diff
@@ -1589,3 +1591,54 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
 
 def align_tensors(*tensors):
     raise RuntimeError('`align_tensors` not yet implemented.')
+
+def result_type(*arrays_and_dtypes):
+    """result_type(*arrays_and_dtypes: Union[Tensor, dtype, Number]) -> dtype
+
+Returns the :class:`torch.dtype` that would result from performing an arithmetic
+operation on the provided input tensors. See type promotion :ref:`documentation <type-promotion-doc>`
+for more information on the type promotion logic.
+
+Args:
+    arrays_and_dtypes (Tensor, Number or dtype): input tensors, numbers or dtypes
+
+Example::
+
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.int), 1.0)
+    torch.float32
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.uint8), torch.tensor(1))
+    torch.uint8
+    >>> torch.result_type(torch.int32, torch.float32)
+    torch.float32
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.int8), 1.0, torch.float32)
+    torch.float32
+"""
+
+    tensors: List[torch.Tensor] = []
+    scalars: List[Number] = []
+    dtypes: List[torch.dtype] = []
+    for x in arrays_and_dtypes:
+        if isinstance(x, Number):
+            scalars.append(x)
+        elif isinstance(x, torch.dtype):
+            dtypes.append(x)
+        elif isinstance(x, torch.Tensor):
+            tensors.append(x)
+        else:
+            raise TypeError(f"result_type(): cannot interpret '{x}' as a data type")
+    if dtypes:
+        dtype = _VF._result_type_dtypes(dtypes)
+        tensors.append(torch.tensor([], dtype=dtype))
+    if scalars:
+        scalar_dtype = _VF._result_type_scalars(scalars)  # type: ignore[arg-type]
+        if tensors:
+            tensor_dtype = _VF.result_type(tensors)  # type: ignore[attr-defined]
+            return _VF.result_type(_VF.tensor([], dtype=tensor_dtype),  # type: ignore[attr-defined]
```

Comment:
At this point it might be preferable to perform an addition like in the reference implementation